stl/inc/variant

```diff
@@ -1447,14 +1448,32 @@ _NODISCARD constexpr bool operator>=(const variant<_Types...>& _Left, const vari
     conjunction_v<is_nothrow_invocable_r<bool, greater_equal<>, const _Types&, const _Types&>...>) /* strengthened */ {
     // determine if _Left's index() is greater than _Right's, or equal and
     // _Left contains a value greater than or equal to _Right
-    using _Visitor             = _Variant_relop_visitor<greater_equal<>, _Types...>;
+    using _Visitor             = _Variant_relop_visitor2<greater_equal<>, bool, _Types...>;
     const size_t _Left_offset  = _Left.index() + 1;
     const size_t _Right_offset = _Right.index() + 1;
     return _Left_offset > _Right_offset
         || (_Left_offset == _Right_offset
             && _Variant_raw_visit(_Right_offset - 1, _Right._Storage(), _Visitor{_Left._Storage()}));
 }
 
+#ifdef __cpp_lib_concepts
+template <class... _Types>
+_NODISCARD constexpr common_comparison_category_t<compare_three_way_result_t<_Types>...>
+    operator<=>(const variant<_Types...>& _Left, const variant<_Types...>& _Right) noexcept(
+        conjunction_v<is_nothrow_invocable_r<common_comparison_category_t<compare_three_way_result_t<_Types>...>,
+            compare_three_way, const _Types&, const _Types&>...>) /* strengthened */ {
+    // Determine the three-way comparison of _Left's and _Right's index, if equal
+    // return the three-way comparison of the contained values of _Left and _Right
+    using _Visitor             = _Variant_relop_visitor2<compare_three_way,
+        common_comparison_category_t<compare_three_way_result_t<_Types>...>, _Types...>;
+    const size_t _Left_offset  = _Left.index() + 1;

```

Comment:
Preexisting: What it the value of the +1 and -1 below?

I see that his is only used for relational operators but I cannot see any value it adds

Reply:
I believe it is to handle the valueless_by_exception case of variant::index() which will return -1 (variant_npos).

Reply:
In the last big `variant` rework I realized that much of the code could be simplified if the index space were contiguous instead of `[0, n) U {~size_t{0}}` and the valueless state treated like just another state for visitors to handle rather than something special to check for and handle before visiting. Adding 1 to the actual indices maps the range of valid values to `[0, n+1)`, thanks to unsigned wraparound, which is nicely contiguous.

For the relational operators, valueless variants are required to be ordered before all variants that hold values and variants with different active alternatives are ordered by `index()`. With the old formulation, we had to check specifically for variants with index `~size_t{0}` (valueless variants) which would otherwise naturally be ordered after variants with `index()` in `[0, n)`. Adding one to the indices trivially puts valueless variants first.

Reply:
For example, the working draft specifies `<=>` as:
```c++
if (v.valueless_by_exception() && w.valueless_by_exception())
  return strong_ordering::equal;
if (v.valueless_by_exception()) return strong_ordering::less;
if (w.valueless_by_exception()) return strong_ordering::greater;
if (auto c = v.index() <=> w.index(); c != 0) return c;
return get<i>(v) <=> get<i>(w);
```
which is equivalent to:
```c++
if (auto c = v.index() + 1 <=> w.index() + 1; c != 0) return c;
return get<i>(v) <=> get<i>(w);
```

Reply:
That makes total sense now that you explained it